src/App/Metadata.cpp

```diff
@@ -0,0 +1,750 @@
+            contentRootElement->appendChild(contentElement);
+            content.second.appendToElement(contentElement);
+        }
+    }
+}
+
+
+void Metadata::parseVersion1(const DOMNode* startNode)
+{
+    auto children = startNode->getChildNodes();
+
+    for (XMLSize_t i = 0; i < children->getLength(); ++i) {
+        auto child = children->item(i);
+        auto element = dynamic_cast<const DOMElement*>(child);
+        if (!element)
+            continue;
+
+        auto tag = element->getNodeName();
+        auto tagString = StrXUTF8(tag).str;
+
+        if (tagString == "name")
+            _name = StrXUTF8(element->getTextContent()).str;
+        else if (tagString == "version")
+            _version = Meta::Version(StrXUTF8(element->getTextContent()).str);
+        else if (tagString == "description")
+            _description = StrXUTF8(element->getTextContent()).str;
+        else if (tagString == "maintainer")
+            _maintainer.emplace_back(element);
+        else if (tagString == "license")
+            _license.emplace_back(element);
+        else if (tagString == "freecadmin")
+            _freecadmin = Meta::Version(StrXUTF8(element->getTextContent()).str);
+        else if (tagString == "freecadmax")
+            _freecadmax = Meta::Version(StrXUTF8(element->getTextContent()).str);
+        else if (tagString == "url")
+            _url.emplace_back(element);
+        else if (tagString == "author")
+            _author.emplace_back(element);
+        else if (tagString == "depend")
+            _depend.emplace_back(element);
+        else if (tagString == "conflict")
+            _conflict.emplace_back(element);
+        else if (tagString == "replace")
+            _replace.emplace_back(element);
+        else if (tagString == "tag")
+            _tag.emplace_back(StrXUTF8(element->getTextContent()).str);
+        else if (tagString == "file")
+            _file.emplace_back(StrXUTF8(element->getTextContent()).str);
+        else if (tagString == "classname")
+            _classname = StrXUTF8(element->getTextContent()).str;
+        else if (tagString == "icon")
+            _icon = fs::path(StrXUTF8(element->getTextContent()).str);
+        else if (tagString == "content")
+            parseContentNodeVersion1(element); // Recursive call
+        else {
+            // If none of this node's children have children of their own, it is a simple element and we
+            // can handle it as a GenericMetadata object
+            auto children = element->getChildNodes();
+            bool hasGrandchildren = false;
+            for (XMLSize_t i = 0; i < children->getLength() && !hasGrandchildren; ++i)
+                if (children->item(i)->getChildNodes()->getLength() > 0)
+                    hasGrandchildren = true;
+            if (!hasGrandchildren)
+                _genericMetadata.insert(std::make_pair(tagString, Meta::GenericMetadata(element)));
+        }
```

Comment:
Maybe a switch..case would be more readable. Also all cases that just 'emplace_back(element)' could be grouped.

Reply:
C++ can't switch on non-integral types -- the strings are listed in the order of their documentation, rather than in the order of their implementation strategy. I don't feel strongly about that order, but it does make sense to me.

Reply:
Correct. And using a constexpr would probably be another mess here. :)